### ISSUE TITLE
[Bug] No attention_mask feed into the model when varlen=False

### ISSUE THREAD

**Issue by alexxchen** [fla-org/flame#41]
### Describe the Bug

It seems that the tokens from different examples will be concatenated into one sequence if varlen=False, and the attention_mask is not feed into the model. This will give totally different from varlen=True. 

### Steps to Reproduce the Bug

From there, the OnlineTokenizedIterableDataset will keep concatenate the examples
```
    def __iter__(self):
        if self.states is not None:
            self.data.load_state_dict(self.states)

        while True:
            for sample in self.tokenize(self.data):
                # keep appending the samples to the token buffer
                self.tokens += sample

                while len(self.tokens) >= self.seq_len:
                    input_ids = torch.tensor(self.tokens[:self.seq_len], dtype=torch.long)
                    self.tokens = self.tokens[self.seq_len:]
                    yield {'input_ids': input_ids}
```
And in DataCollatorForLanguageModeling, the the padding is just accroding to the length of the example processed from OnlineTokenizedIterableDataset
```
        if not self.varlen:
            # --- Handling for varlen=False (Batch Padding) ---
            length_of_first = examples[0]['input_ids'].size(0)
            needs_padding = not all(example['input_ids'].size(0) == length_of_first for example in examples)

            if needs_padding:
                # Check for pad token if padding is actually required
                if self.tokenizer.pad_token_id is None:
                    raise ValueError(
                        f'You are attempting to pad samples but the tokenizer you are using '
                        f'({self.tokenizer.__class__.__name__}) does not have a pad token.'
                    )
                # Pad using the tokenizer, ensuring attention_mask is returned
                batch = self.tokenizer.pad(examples, return_tensors='pt', return_attention_mask=True)
            else:
                # No padding needed, stack directly and create a full attention mask
                input_ids = torch.stack([example['input_ids'] for example in examples], dim=0)
                batch = {
                    'input_ids': input_ids,
                    # Create attention mask of all ones
                    'attention_mask': torch.ones_like(input_ids),
                }
```
As a result, the attention_mask will be all ones when varlen=False

### Expected Behavior

The tokens from different examples should not be concatenate into one sequence when varlen=False.
And attention_mask should be [0,0,0,0,0,1,1,1,1] when using left padding.

### Environment Information

1. Torch:2.6.0
2. Triton:>3.0

**Comment by yzhangcs** [comment]
> It seems that the tokens from different examples will be concatenated into one sequence

Hi, isn’t this the standard practice in earlier LLM training pipelines (e.g., Llama) when varlen is disabled?

**Comment by alexxchen** [comment]
Hi, thanks for you reply! 
In earlier LLM training pipelines, the simplest approach was to pad either the left or right side of a single example, while using an attention mask to prevent attention values from being computed for the padded positions. If two unrelated pieces of text are directly concatenated without providing an attention mask, then the two completely unrelated sequences will indeed attend to each other, correct?

**Comment by yzhangcs** [comment]
I personally also recommend using varlen during the 4k/8k pretraining stage which is faster as the inputs are composed of several short pieces.
